Pub/Sub: add wrangler pubsub commands

elithrar opened the pull request:
This PR adds support for Pub/Sub (https://developers.cloudflare.com/pub-sub/) via a new `wrangler pubsub` command group.

- [x] Finalize MVP tests
- [x] Add beta caveat / notice (?) ala pages - e.g. `Pub/Sub is in private beta. See https://developers.cloudflare.com/pub-sub for details on how to get access to Pub/Sub`
- [x] Bikeshed over API naming conventions
- [x] @oliy to sign off as ~co-conspirator~ eng lead for Pub/Sub.

elithrar commented:
Rebased against master and fixed _all_ of the merge conflicts.

elithrar commented:
@rozenmd - done:

- `index.tsx` now imports `pubsubCommands`: `export function pubSubCommands(pubsubYargs: Argv, subHelp: CommandModule): Argv`
- All the commands are isolated in `pubsub-commands.tsx`
- API functions remain in `pubsub.ts`

oliy commented on packages/wrangler/src/pubsub-commands.tsx:
```diff
@@ -0,0 +1,625 @@
+
+              console.log(resp);
+
+              // logger.log(
```
Should probably remove this now.

elithrar replied:
Agree, fixing...

oliy commented on packages/wrangler/src/pubsub.ts:
```diff
@@ -0,0 +1,281 @@
+): Promise<Record<string, string>> {
+  let url = `/accounts/${accountId}/pubsub/namespaces/${namespace}/brokers/${broker}/credentials?number=${number}&type=${type}`;
+  if (clientIds) {
+    const ids = makeQueryString(clientIds, "clientid");
```
This works, but might want to just use URLSearchParams in the future, since it automatically supports proper encoding, multiple values, etc.

elithrar replied:
I'll fix this. The original code relied on the `URL` type, which isn't valid as there's no baseURL set (since it's set at a package level).

The "gap" is that the best you can do is concat `path + "?" + params.toString()`.

elithrar replied:
I've fixed this to use `URLSearchParams` - but we still have to concat the params.

```ts
  let url = `/accounts/${accountId}/pubsub/namespaces/${namespace}/brokers/${broker}/credentials`;

  const params = new URLSearchParams();
  params.append("number", `${number}`);
  params.append("type", type);

  if (clientIds) {
    for (const id of clientIds) {
      params.append("clientid", id);
    }
  }

  if (expiration) {
    params.append("expiration", `${expiration}`);
  }

  // We have to concat these as the URL class requires a base URL, which we
  // don't know within the scope of this function.
  url = url + `?${params.toString()}`;

  return await fetchResult<Record<string, string>>(url);
  ```

elithrar commented on packages/wrangler/src/pubsub.ts:
```diff
@@ -0,0 +1,281 @@
+  accountId: string,
+  namespace: PubSubNamespace
+): Promise<void> {
+  return await fetchResult<void>(`/accounts/${accountId}/pubsub/namespaces`, {
```
This passes the Prettier config for the repo (e.g. all files are formatted) so changing it will fail the checks.

elithrar replied:
Manually tweaked:

```ts
  return await fetchResult<void>(
    `/accounts/${accountId}/pubsub/namespaces`,
    { 
      method: "POST",
      body: JSON.stringify(namespace),
    }
  );
```

... with Prettier format run (back to same as review):

```ts
  return await fetchResult<void>(
    `/accounts/${accountId}/pubsub/namespaces`,
    { 
      method: "POST",
      body: JSON.stringify(namespace),
    }
  );
```

elithrar commented:
Addressed comments. Tests continue to pass.

@rozenmd / @threepointone / @petebacondarwin  - this is ready for review and (ideally) merge.

If we can also get a new Wrangler point release that includes this new surface it'd be great: we want to make sure our docs recommend users "install or upgrade to Wrangler version <X.y.z.> to use the new `pubsub` commands" so that they aren't wondering why their version doesn't have them.

threepointone commented:
If you run `cd packages/wrangler && npm test -- -u`, it'll update the snapshots and pass tests, happy to review, stamp and merge right after!

threepointone commented:
It looks like you've committed a folder from a previous version of the repo, specifically the `examples` folder. Delete that?

threepointone commented:
I'm going to try pushing commits to your branch to see if I can help out here.

threepointone reviewed (approved):
This looks great! Let's land it and take it for a spin, and we can iterate further.

threepointone commented:
Shipped! Available in wrangler@beta, and will be included in the next release 2.0.16

elithrar commented:
Thanks @threepointone!
